### PR TITLE
[libc++] Make benchmarks forward-compatible with the test suite

### PR DESCRIPTION
**Pull request by ldionne**
This patch fixes warnings and errors that come up when running the benchmarks as part of the test suite. It also adds the necessary Lit annotations to make it pass in various configurations and increases the portability of the benchmarks.

### CODE REVIEW

File: libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp [hunk 1]
```diff
@@ -19,7 +21,8 @@ static void bm_lexicographical_compare(benchmark::State& state) {
   for (auto _ : state) {
     benchmark::DoNotOptimize(vec1);
     benchmark::DoNotOptimize(vec2);
-    benchmark::DoNotOptimize(std::lexicographical_compare(vec1.begin(), vec1.end(), vec2.begin(), vec2.end()));
+    bool result = std::lexicographical_compare(vec1.begin(), vec1.end(), vec2.begin(), vec2.end());
+    benchmark::DoNotOptimize(result);
```
Comment (philnik777) [inline libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp:25]: What is the problem here?

File: libcxx/test/benchmarks/ordered_set.bench.cpp [hunk 2]
```diff
@@ -181,7 +184,7 @@ struct IterateRangeFor : Base {
     while (State.KeepRunningBatch(TableSize * NumTables)) {
       for (auto& Set : Data.Sets) {
         for (auto& V : Set) {
-          benchmark::DoNotOptimize(V);
+          benchmark::DoNotOptimize(const_cast<std::set<uint64_t>::reference>(V));
```
Comment (philnik777) [inline libcxx/test/benchmarks/ordered_set.bench.cpp:187]: Why did this change?

File: libcxx/test/benchmarks/util_smartptr.bench.cpp [hunk 3]
```diff
@@ -13,14 +13,16 @@
 static void BM_SharedPtrCreateDestroy(benchmark::State& st) {
   while (st.KeepRunning()) {
     auto sp = std::make_shared<int>(42);
-    benchmark::DoNotOptimize(sp.get());
+    int* ptr = sp.get();
+    benchmark::DoNotOptimize(ptr);
```
Comment (philnik777) [inline libcxx/test/benchmarks/util_smartptr.bench.cpp:17]: Same question here.

File: libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp:25]: `DoNotOptimize` shouldn't be called with a temporary. A warning is emitted, but before running this stuff as part of the test suite, we just ignore that warning. After, it gets turned into an error.

File: libcxx/test/benchmarks/ordered_set.bench.cpp (hunk 2 above)
Comment (ldionne) [reply libcxx/test/benchmarks/ordered_set.bench.cpp:187]: Same here as above: `error: 'DoNotOptimize<unsigned long long>' is deprecated: The const-ref version of this method can permit undesired compiler optimizations in benchmarks [-Werror,-Wdeprecated-declarations]`

File: libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp (hunk 1 above)
Comment (philnik777) [reply libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp:25]: @ldionne The problem isn't the temporary. It's `const` values being passed, at least according to https://github.com/google/benchmark/pull/1493. This has been fixed by https://github.com/google/benchmark/pull/1608 in v1.8.1. Are you possibly using v1.8.0?

Comment (ldionne) [reply libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp:25]: The version of GoogleBenchmarks used here is the one in the monorepo under `third-party/benchmark`. It looks like that is 1.8.3 right now based on `third-party/benchmark/MODULE.bazel`.

Comment (ldionne) [reply libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp:25]: Removing this diff seems not to make any difference. I'll remove it.

Comment (ldionne) [reply libcxx/test/benchmarks/algorithms/lexicographical_compare.bench.cpp]: I suspect what happened is that I started this patch before GoogleBenchmark was updated to 1.8.3 in the monorepo. I went through the patch and removed the changes that weren't necessary anymore related to temporaries.

**Review by philnik777 — APPROVED** [review]
LGTM as a first step. I think we'll want to do some more cleanup later to make them more portable though. @StephanTLavavej do you have any interest in sharing benchmarks across libc++/the STL?

**Comment by StephanTLavavej** [comment]
Sure, that’s interesting! We’ve been building up a set of benchmarks using Google Benchmark.

**Pull request merged**